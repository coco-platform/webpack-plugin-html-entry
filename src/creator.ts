/**
 * @description - create html element literal
 * @author - huang.jian <hjj491229492@hotmail.com>
 */

export function createScript(pathname: string): string {
  return `<script src="${pathname}"></script>`;
}

export function createStylesheet(pathname: string): string {
  return `<link rel="stylesheet" href="${pathname}" />`;
}

export function createRoot(): string {
  return '<div class="root"></div>';
}
