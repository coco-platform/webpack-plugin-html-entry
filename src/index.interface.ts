export interface HTMLEntryPluginOptions {
  filename?: string;
  extraHTML?: string | (() => string);
}
