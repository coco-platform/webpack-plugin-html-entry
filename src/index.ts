/* eslint-disable import/prefer-default-export */
/* eslint-disable indent */
/* eslint-disable class-methods-use-this */

// package
import url from 'url';
import { Compiler, Compilation, sources, WebpackPluginInstance } from 'webpack';

// internal
import { createExtraHTML, createScript, createStylesheet } from './creator';
import { HTMLEntryPluginOptions } from './index.interface';

// scope
const { RawSource } = sources;

const defaultOptions: HTMLEntryPluginOptions = {
  filename: 'index.html',
  extraHTML: createExtraHTML,
};

export class HTMLEntryPlugin implements WebpackPluginInstance {
  // padding optional property with default options
  private readonly options: Required<HTMLEntryPluginOptions>;

  constructor(options: HTMLEntryPluginOptions = {}) {
    this.options = { ...defaultOptions, ...options } as Required<
      HTMLEntryPluginOptions
    >;
  }

  apply(compiler: Compiler): void {
    compiler.hooks.compilation.tap('HTMLEntry', (compilation) => {
      compilation.hooks.processAssets.tap(
        {
          name: 'HTMLEntry',
          stage: Compilation.PROCESS_ASSETS_STAGE_ADDITIONAL,
        },
        () => {
          const entrypoint = compilation.entrypoints.get('main');

          if (!entrypoint) {
            throw new Error(
              // eslint-disable-next-line quotes
              `HTMLEntryPlugin can't recognize entrypoints, only support single 'main' entrypoint`
            );
          }

          // runtime chunk first
          // TODO - resolve complicated dependency graph
          const chunks = entrypoint.chunks.sort((a) =>
            a.hasRuntime() ? -1 : 1
          );
          const files = chunks.reduce<string[]>((acc, chunk) => {
            return [...acc, ...chunk.files];
          }, []);

          // element order:
          //   1. stylesheet
          //   2. script
          //   3. html inline
          const source = files
            .sort((a) => (a.endsWith('.css') ? -1 : 1))
            .map((filename) =>
              url.resolve(
                compiler.options.output.publicPath as string,
                filename
              )
            )
            .map((filename) => {
              switch (true) {
                case filename.endsWith('.css'):
                  return createStylesheet(filename);
                case filename.endsWith('.js'):
                  return createScript(filename);
                default:
                  // just ignore lint waning
                  return undefined;
              }
            })
            // extra html as qiankun microservice container
            .concat(
              typeof this.options.extraHTML === 'string'
                ? this.options.extraHTML
                : this.options.extraHTML()
            )
            .join('\n');

          compilation.emitAsset(this.options.filename, new RawSource(source));
        }
      );
    });
  }
}
