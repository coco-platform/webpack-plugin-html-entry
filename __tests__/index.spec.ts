/**
 * @description - suits example
 * @author - huang.jian <hjj491229492@hotmail.com>
 */

/* eslint-disable import/no-extraneous-dependencies, no-console */

// package
import path from 'path';
import webpack, { Compiler } from 'webpack';
import { fs as memfs } from 'memfs';

// Internal
import { HTMLEntryPlugin } from '../src/index';

describe('plugin test suits', () => {
  it('should complete standard workflow', (done) => {
    const configuration = {
      entry: path.resolve(__dirname, '../__tests__/__fixture__/index.js'),
      output: {
        path: path.resolve(__dirname, '../dist'),
        filename: '[name].js',
        publicPath: '/',
        library: 'html_entry',
      },
      plugins: [
        new HTMLEntryPlugin({
          filename: 'index.entry.html',
        }),
      ],
    };
    const compiler = webpack(configuration as webpack.Configuration);
    const outputPath = `${configuration.output.path}/index.entry.html`;

    compiler.outputFileSystem = memfs as Compiler['outputFileSystem'];
    compiler.run((err, stats) => {
      try {
        expect(err).toBeNull();
        // ensure stats not undefined
        expect(stats).toBeTruthy();
        // use optional symbol avoid lint rules
        expect(stats?.hasErrors()).toBeFalsy();
        expect(memfs.readFileSync(outputPath, 'utf8')).toMatchSnapshot();
        done();
      } catch (error) {
        done(error);
      }
    });
  });

  it('should support customize extra html', (done) => {
    const configuration = {
      entry: path.resolve(__dirname, '../__tests__/__fixture__/index.js'),
      output: {
        path: path.resolve(__dirname, '../dist'),
        filename: '[name].js',
        publicPath: '/',
        library: 'html_entry',
      },
      plugins: [
        new HTMLEntryPlugin({
          filename: 'index.html',
          extraHTML: '<section class="ms-container"></section>',
        }),
      ],
    };
    const compiler = webpack(configuration as webpack.Configuration);
    const outputPath = `${configuration.output.path}/index.html`;

    compiler.outputFileSystem = memfs as Compiler['outputFileSystem'];
    compiler.run((err, stats) => {
      try {
        expect(err).toBeNull();
        // ensure stats not undefined
        expect(stats).toBeTruthy();
        // use optional symbol avoid lint rules
        expect(stats?.hasErrors()).toBeFalsy();
        expect(memfs.readFileSync(outputPath, 'utf8')).toMatchSnapshot();
        done();
      } catch (error) {
        done(error);
      }
    });
  });
});
