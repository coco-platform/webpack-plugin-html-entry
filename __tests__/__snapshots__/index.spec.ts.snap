// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugin test suits should complete standard workflow 1`] = `
"<script src=\\"/main.js\\"></script>
<div class=\\"root\\"></div>"
`;
